File: dp/minimizingcoins.cc
#include <bits/stdc++.h>
using namespace std;

int main()
{
    int n, x;
    cin >> n >> x;
    int li[n];
    for (int i = 0; i < n; i++)
        cin >> li[i];
    int infi = 10000000;
    vector<int> dp(x + 1, infi);
    dp[0] = 0;

    for (int j = 1; j <= x; j++)
    {

        int ans = infi;

        for (int k = 0; k < n; k++)
        {

            if (li[k] <= j)
            {
                ans = min(ans, dp[j - li[k]]);
            }
        }

        if (ans == infi)
        {
            dp[j] = infi;
        }
        else
        {
            dp[j] = ans + 1;
        }
    }
    if (dp[x] == infi)
        cout << -1;
    else
        cout << dp[x];
    return 0;
}

File: dp/minimizingcoins.py
from sys import stdin

n, amount = map(int, stdin.readline().split())
coins = [int(c) for c in stdin.readline().split()]

infi = 10 ** 9
dp = [infi] * (amount + 1)
dp[0] = 0

for i in range(1, amount + 1):
    ans = infi
    for j in range(n):
        if coins[j] <= i:
            ans = min(ans, dp[i - coins[j]])
    if ans == infi:
        dp[i] = infi
    else:
        dp[i] = ans + 1
if dp[amount] == infi:
    print(-1)
else:
    print(dp[amount])
